UIUX: Overhaul of UTXO list, handling of locked UTXOs and scrollbar added to tx-table web component

moneymanolis opened the pull request:
### Main
- Ensure that locked UTXO / UTXO used in unsigned transactions can't be used for transactions (fixes #1562)
- Overhaul of checkbox logic in UTXO list
      1. If a locked or "locked-unsigned" UTXO is chosen, only UTXO of the respective type  can be selected anymore.
      2. Mouseover event for checkboxes added (flashing "Select checkboxes for actions.")
      3. Manage PSBT button action added
      4. Icon info popups added
See the following video:
https://user-images.githubusercontent.com/70536101/154949485-b7f45060-15f9-46a3-8b53-18268583720c.mov

- Locked UTXO or UTXO in unsigned transactions can't be chosen in the coin selection menu anymore, the checkbox is replaced by the snowflake image and the info text "Unsigned" is added to distinguish between locked and "unsigned-locked" UTXO:

<img width="422" alt="image" src="https://user-images.githubusercontent.com/70536101/154949077-f0758eba-77fb-42cb-b087-203623ead52f.png">

### Misc
- Fix rounding error in coin selection (fixes #1568)
- Fix styling / text in general settings.
- Design changes to make the UTXO- / TX-list display quieter.

moneymanolis commented:
@relativisticelectron if you want, you could test the PR to check whether you are happy with the overall UX. Would be great. I am happy to adapt it.

relativisticelectron commented:
Hi, I think the freezing symbol is cute.  
I found however a way to still create multiple conflicting transactions:

- Create tx with a utxo
- Go to utxo overview.  Unfreeze the utxo.
- Create a tx with the utxo again

I am actually not sure if this should be prevented (As I said in another issue, I  actually want to create conflicting tx with differents fees).  I just wanted to point it out, and let you decide if you want to close this loophole or not.

moneymanolis commented:
Good point, I think we should add a little text in the front end to distinguish between UTXOs that are locked because they are used in an unsigned tx and UTXO that are locked by manually freezing them. Server sidewise, Specter lockes both via the `lockunspent` RPC call.

moneymanolis commented:
@relativisticelectron I reworked the UX quite a bit and also incorporated your latest point. If you could give it a try and provide feedback, that would be awesome!

relativisticelectron commented:
I really like it.  Good idea with the different buttons that conditionally appear. And even the checkboxes are conditionally now (cool logic!)
![image](https://user-images.githubusercontent.com/60378539/154957645-f97ee3fc-8c20-4b7e-9c7d-8a16376411a0.png)

The only way to achieve an inconsistency (with the freezing status) was the following:
- Create tx that spends utxo0
- Import psbt that also spends this utxo0
- Deleting one of the transactions again   (then it can occur that the status of utxo0 is either locked, unlocked, or still frozen)

This is however such an edge case, that I am not sure it is worth it catching it.

moneymanolis commented:
> The only way to achieve an inconsistency (with the freezing status) was the following:
> 
>     * Create tx that spends utxo0
> 
>     * Import psbt that also spends this utxo0
> 
>     * Deleting one of the transactions again   (then it can occur that the status of utxo0 is either locked, unlocked, or still frozen)
> 
> 
> This is however such an edge case, that I am not sure it is worth it catching it. And if specter should catch it, it would be in the import logic of psbts; so it probably should not be part of this PR, but a separate one (if at all).

@relativisticelectron I will address this in a separate PR, already created an issue for it here:
https://github.com/cryptoadvance/specter-desktop/issues/1582
Thank you for taking the time and providing feedback!

relativisticelectron commented:
> @relativisticelectron I will address this in a separate PR, already created an issue for it here:
> #1582
> Thank you for taking the time and providing feedback!

Thank you for doing all this great work!

relativisticelectron commented:
I noticed after more testing, that sometimes the utxo appear "locked" and sometimes "unsigned". I don't know why.  There is 1 psbt consuming all these utxos:
![image](https://user-images.githubusercontent.com/60378539/155591120-b1a22180-7b75-4624-a8e5-302c9a9c8706.png)

moneymanolis commented:
Good catch, I will have a look, thanks.

relativisticelectron commented on src/cryptoadvance/specter/templates/includes/tx-table.html:
```diff
@@ -1,9 +1,48 @@
+    scrollbar-default="off"
+    compact-switch="true">
+</tx-table> 
+-->
```
`#}`  end jinja2 comment.

moneymanolis replied:
Added in https://github.com/cryptoadvance/specter-desktop/pull/1580/commits/05c5a62e0bac83eab31dfda5d2d935b3bb67adef

relativisticelectron commented on src/cryptoadvance/specter/templates/includes/tx-table.html:
```diff
@@ -1,9 +1,48 @@
+<!-- 
```
Please add `{#`  here, to indicate for jinja2 it should not replace here something.  Otherwise I get the error

```
  File "/home/<user>/programs/anaconda3/lib/python3.9/site-packages/cryptoadvance/specter/templates/includes/tx-table.html", line 35, in top-level template code
    wallet="{{ wallet.alias }}"
  File "/home/<user>/programs/anaconda3/lib/python3.9/site-packages/jinja2/environment.py", line 471, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'wallet' is undefined
```

relativisticelectron commented on src/cryptoadvance/specter/templates/wallet/history/components/total_wallet_balances.jinja:
```diff
@@ -9,23 +9,24 @@
-                     {% endif %}
-            )</small>
+        {% if wallet.balance.get("untrusted_pending", 0) or wallet.balance.get("immature", 0) %}
+        <div class="smaller-card">
```
![image](https://user-images.githubusercontent.com/60378539/157185870-e995e0d9-10c3-4197-837c-4a09a5af081b.png)
It would be even nicer if the amounts would be _left_ aligned, so the decimal separator is at the same x position.   And for completeness the unit  (tBTC) after each amount would be good too.

moneymanolis replied:
Good points, I've created a separate issue for that since this was not really meant to be in this PR in the first place.

relativisticelectron commented on src/cryptoadvance/specter/static/styles.css:
```diff
@@ -702,29 +707,67 @@ tbody tr:hover{
+
+main {
+  overflow-y: scroll;
+  scrollbar-color: hsl(40 100% 75%) hsl(40 100% 50%);
```
The chosen colors are not appealing.  How about something like  
![image](https://user-images.githubusercontent.com/60378539/157189533-4db82a2e-b84a-4f83-8aa4-698fe963e75d.png)
it looks like
![image](https://user-images.githubusercontent.com/60378539/157189609-e767f185-4e3e-42e3-b91d-6a98026e5839.png)

moneymanolis replied:
How can you not like orange :-). I wanted to have two distinct different colors for the two scrollbars. Let's do it like this I will discuss it with our design guy and present your suggestion.

moneymanolis replied:
Addressed in https://github.com/cryptoadvance/specter-desktop/pull/1580/commits/05c5a62e0bac83eab31dfda5d2d935b3bb67adef

relativisticelectron commented on src/cryptoadvance/specter/templates/includes/tx-table.html:
```diff
@@ -91,7 +125,87 @@
+            min-height: 350px;
+            overflow-y: scroll;
+            scrollbar-color: grey hsl(217, 15%, 55%); 
+            scrollbar-width: thin; 
```
How about
![image](https://user-images.githubusercontent.com/60378539/157190982-ffcccfdb-ee20-481a-88d4-bd5dc5dee673.png)
looking like
![image](https://user-images.githubusercontent.com/60378539/157190941-b10035f7-be5f-4553-a3fc-318cba415b48.png)

relativisticelectron reviewed (commented):
Thanks so much for the great work. I could not create inconsistencies.

moneymanolis commented on src/cryptoadvance/specter/wallet.py:
```diff
@@ -1544,21 +1544,21 @@ def update_balance(self):
-            available["untrusted_pending"] = round(available["untrusted_pending"], 8)
-            balance["trusted"] = available["trusted"]
-            balance["untrusted_pending"] = available["untrusted_pending"]
+            # locked_utxo = self.rpc.listlockunspent()
```
Changed in https://github.com/cryptoadvance/specter-desktop/pull/1580/commits/47da7a86330d3f0ef8f53ec559de682950574f44

moneymanolis commented:
@relativisticelectron see my latest commit: This is a compromise regarding the scrollbar functionality and design also after some internal discussions. Thanks for your input.

moneymanolis commented:
@cypherhoodlum would be great to get feedback from you on this PR regarding the scrollbar implementation for the main div, the tx table and the tx table used in the coin selection. (Of course, if you stumble upon anything else which you would improve UIUX-wise in this PR, let me know as well)

cypherhoodlum commented:
> @cypherhoodlum would be great to get feedback from you on this PR regarding the scrollbar implementation for the main div, the tx table and the tx table used in the coin selection. (Of course, if you stumble upon anything else which you would improve UIUX-wise in this PR, let me know as well)

I think the scrollbars look and feel great. No issues there.

However, I think the new way of showing the balances is slightly harder to read than the original version.

Original:
![Screenshot from 2022-03-23 00-15-25](https://user-images.githubusercontent.com/82510823/159585935-811efec8-ebd0-46dd-b869-a87bc69e3971.png)


New:
![Screenshot from 2022-03-23 00-15-59](https://user-images.githubusercontent.com/82510823/159585962-87d37052-6467-4fac-8e5d-138b4ea50b98.png)

I think it's good to unify the overall visual look by putting elements in boxes with rounded borders (new version) but the readability should not suffer. Maybe the first row could have a bigger font size than the others? Another thing worth trying out is to align the numbers/balances vertically. Maybe the whole box could be a bit bigger since it's probably the most important element on the whole page. It's at least the second most important element if the transaction history is the most important one. So I think the whole box and the total amount could be accentuated some more. What do you think @moneymanolis ?

cypherhoodlum commented:
Maybe something like this:

<img src="https://user-images.githubusercontent.com/82510823/159592700-db61eef9-eef6-4edf-9c63-55ebb7fad89a.png" width="600" />

moneymanolis commented:
> I think it's good to unify the overall visual look by putting elements in boxes with rounded borders (new version) but the readability should not suffer. Maybe the first row could have a bigger font size than the others? Another thing worth trying out is to align the numbers/balances vertically. Maybe the whole box could be a bit bigger since it's probably the most important element on the whole page. It's at least the second most important element if the transaction history is the most important one. So I think the whole box and the total amount could be accentuated some more. What do you think @moneymanolis ?

Thanks a lot for the feedback and the PR, @cypherhoodlum.
Good points! 

Since I would like to wrap this PR up soon (it is already too big actually), though, I would keep the style for now and move the discussion and work on the balance box to this already opened issue: #1627

k9ert commented:
@cypherhoodlum awesome feedback and thanks for your PR. It's yet a big improvement. But i also agree with @moneymanolis , let's not make that PR any bigger. We struggled now a lot with the UI-tests in this PR. The last 12 commits are only regarding improving testing. It's high likely your PR would break them (again). And even if not, this PR is a moving goal-post since almost the beginning.
There are people who think that feature-branches should only be open for a couple of hours. This one is now open for more than 40 days.

cypherhoodlum commented:
> But i also agree with @moneymanolis , let's not make that PR any bigger. We struggled now a lot with the UI-tests in this PR. The last 12 commits are only regarding improving testing. It's high likely your PR would break them (again).

Right, that's understandable. I will make a separate PR related to #1627.

k9ert commented:
This is really an unlucky PR. After scratching our heads why the builds are not kicking in at all (probably an [issue](https://status.cloud.google.com/) with the Google-cloud), now the linter is broken.
```
Traceback (most recent call last):
  File "/usr/local/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "src/black/__init__.py", line 1423, in patched_main
  File "src/black/__init__.py", line 1409, in patch_click
ImportError: cannot import name '_unicodefun' from 'click' (/usr/local/lib/python3.10/site-packages/click/__init__.py)
```
Investigating now ...

k9ert commented:
Fixed in https://github.com/cryptoadvance/specter-desktop/pull/1642
Build is now hopefully green.
